### PR TITLE
Feature/optimise dockerfile

### PR DESCRIPTION
**Pull request by martyn-w**
* cache gems centrally so they can be reused by both geoblacklight and
willow
* simplify and streamline Dockerfiles
* remove superfluous statements
* update rails from 5.0.1 to 5.0.2 to match geoblacklight (and share same gems)

### CODE REVIEW

**Comment by martyn-w** [comment]
NB: you will need to drop your existing database volumes inorder to run this; ie:
`docker-compose down --volumes && docker-compose build && docker-compose up`

**Comment by emanuil-tolev** [comment]
PR meant to improve Dockerfiles and the *development* use of Docker.

File: willow/Dockerfile [hunk 1]
```diff
@@ -34,39 +29,19 @@ RUN apt-get update -qq && \
...
-		echo '#!/bin/sh'; \
-		echo 'set -e'; \
-		echo; \
-		echo 'dirname "$(dirname "$(readlink -f "$(which javac || which java)")")"'; \
-	} > /usr/local/bin/docker-java-home && \
-	chmod +x /usr/local/bin/docker-java-home && \
-    set -x && [ "$JAVA_HOME" = "$(docker-java-home)" ]
+	&& rm -rf /var/lib/apt/lists/* \
+	&& /var/lib/dpkg/info/ca-certificates-java.postinst configure \
+    && mkdir -p /fits/ \
```
Comment (emanuil-tolev) [inline willow/Dockerfile]: I know we'd like to reduce the number of Docker commands used, but I think we can spare a separate RUN command for installing fits, as opposed to putting it at the end of the apt-get install process.

File: willow/docker-entrypoint.sh [hunk 2]
```diff
@@ -1,24 +1,20 @@
...
 
-echo Creating tmp and log
-mkdir -p tmp/pids log
-rm  -f tmp/pids/*
+# Install any missing gems
+bundle check || bundle install
 
-echo "Mounting willow source code as a Docker volume with symlinks, except certain temporary directories and other files"
-
-find ../willow_source/ -maxdepth 1 ! -name "log" ! -name "tmp" ! -name "docker-entrypoint.sh" -exec ln -sf {} ';'
```
Comment (emanuil-tolev) [inline willow/docker-entrypoint.sh]: So we don't need to do this anymore and "it just works"?

File: willow/Dockerfile (hunk 1 above)
Comment (martyn-w) [inline willow/Dockerfile]: @emanuil-tolev any reason why it should be separate from the other lines?

File: willow/docker-entrypoint.sh (hunk 2 above)
Comment (martyn-w) [inline willow/docker-entrypoint.sh]: yep, try it and see

File: willow/Dockerfile (hunk 1 above)
Comment (emanuil-tolev) [inline willow/Dockerfile]: I see each RUN as a group of commands that is related in some manner. For example, setting up apt repositories, running apt-get update and installing a bunch of system software.

I just saw fits as a separate (unrelated) thing to do.

I also find the pattern of super-long commands kind of smelly, but I understand why it's being used. I have also observed well maintained open source images (postgres, solr) using it in their own Dockerfiles. I'd find putting every single shell command in the Dockerfile in a single RUN directive a bit extreme though.

File: willow/docker-entrypoint.sh (hunk 2 above)
Comment (emanuil-tolev) [inline willow/docker-entrypoint.sh]: Yep, seems to work. Brilliant.

File: willow/Dockerfile (hunk 1 above)
Comment (emanuil-tolev) [inline willow/Dockerfile]: Basically I think it's less readable. We start RUN with "apt-get ..." and you do a whole lot of apt-get stuff. Then a long way down, we start something unrelated (fits).

You don't have to change it, I'm just discussing it :)

Comment (martyn-w) [inline willow/Dockerfile]: Hi @emanuil-tolev I think you are right and Readability is important; I have moved it to a separate line

**Pull request merged**